maven-artifact-notifier-webapp/src/main/java/fr/openwide/maven/artifact/notifier/web/application/administration/page/AdministrationArtifactPortfolioPage.java

```diff
@@ -51,12 +55,33 @@ public AdministrationArtifactPortfolioPage(PageParameters parameters) {
 
 			@Override
 			public void onClick() {
+				final AtomicReference<Exception> exception = new AtomicReference<Exception>(null);
 				try {
-					mavenSynchronizationService.synchronizeAllArtifactsAndNotifyUsers();
-					getSession().success(getString("administration.artifact.syncAll.success"));
-				} catch (Exception e) {
-					LOGGER.error("Error occured while synchronizing artifacts", e);
+					Thread thread = new Thread(new Runnable() {
+						@Override
+						public void run() {
+							try {
+								mavenSynchronizationService.synchronizeAllArtifactsAndNotifyUsers();
+							} catch (ServiceException|SecurityServiceException|InterruptedException e) {
+								exception.set(e);
+								if (e instanceof InterruptedException) {
+									Thread.currentThread().interrupt();
+								}
+							}
+						}
+					});
+					thread.start();
+					thread.join();
```

Comment:
Why do you start it in a new Thread if you need to wait for it here ? Why not just execute it in the current thread ?

Reply:
Synchronization process need to run in a separate Hibernate session for performance reason. Running it into a separate Thread and joining the thread is a way to run it with an isolated session without interrupting the HTTP thread&request-scoped session.

Another way to proceed would be to deal with Thread-scoped/TransactionSynchronizationManager registered session (save, replace, artifact synchronization, restore): it does not seem to me there is an helper for this ?

Reply:
You know better the details how this leads to better performance in your app, but this is how I see it from outside:

Wicket is thread-per-request based so if you use OpenSessionInViewFilter then you have one request scoped DB transaction. If you need to start a new one for the Maven sync then you can use `@Propagation.RequiresNew` for this service method. If you don't make it in a separate transaction then the only drawback I see is that the request-scoped transaction may get rolled back if something fails later.

Reply:
We are transaction-less at the time I do this trick (only session is opened at request scope, transaction are kept at service scope). This pattern is a bit original and can be tricky when rollbacks are involved, but work flawlessly for a lot of our projects.

I do not use @Transactional mechanism as I do not want to wrap whole synchronization in a transaction (we deal with transaction around each artifact synchronization). I do not dig into @Transactional configuration to see if it is possible to let it open a session without starting a new transaction.

As transactions are also separated (Thread trick also isolates transactions), there is no request-scoped transaction, and page / synchronization data do not need to be synced, I do not think there is any problem with transaction isolation / rollback.